src/main/java/io/wisetime/connector/utils/DurationCalculator.java

```diff
@@ -0,0 +1,147 @@
+/*
+ * Copyright (c) 2018 Practice Insight Pty Ltd. All Rights Reserved.
+ */
+
+package io.wisetime.connector.utils;
+
+import java.util.function.Function;
+
+import io.wisetime.generated.connect.TimeGroup;
+import io.wisetime.generated.connect.TimeRow;
+
+/**
+ * Utility to assist with calculating durations for a TimeGroup.
+ *
+ * @author shane.xie@practiceinsight.io
+ */
+public class DurationCalculator {
+
+  private TimeGroup timeGroup;
+  private DurationSource durationSource = DurationSource.TIME_GROUP;
+  private boolean useExperienceRating = true;
+
+  /**
+   * Create a DurationCalculator for a TimeGroup.
+   *
+   * @param timeGroup the TimeGroup whose durations we want to calculate
+   * @return a DurationCalculator with the default configuration equivalent to:
+   *
+   *   DurationCalculator.of(timeGroup)
+   *       .useDurationFrom(DurationSource.TIME_GROUP)
+   *       .useExperienceRating()
+   */
+  public static DurationCalculator of(final TimeGroup timeGroup) {
+    return new DurationCalculator(timeGroup);
+  }
+
+  private DurationCalculator(final TimeGroup timeGroup) {
+    this.timeGroup = timeGroup;
+  }
+
+  /**
+   * Tell the calculator whether to read the duration from the TimeGroup or its TimeRows.
+   *
+   * @param durationSource which DurationSource option to use
+   * @return the DurationCalculator with the duration from option applied
+   */
+  public DurationCalculator useDurationFrom(final DurationSource durationSource) {
+    this.durationSource = durationSource;
+    return this;
+  }
+
+  /**
+   * Tell the calculator to use the user's experience rating in its calculations.
+   *
+   * @return the DurationCalculator configured to use the user's experience rating
+   */
+  public DurationCalculator useExperienceRating() {
+    useExperienceRating = true;
+    return this;
+  }
+
+  /**
+   * Tell the calculator to disregard the user's experience rating in its calculations.
+   *
+   * @return the DurationCalculator configured to ignore the user's experience rating
+   */
+  public DurationCalculator disregardExperienceRating() {
+    useExperienceRating = false;
+    return this;
+  }
+
+  /**
+   * Calculate durations for the TimeGroup.
+   *
+   * @return Result containing the calculated per-tag and total durations
+   */
+  public Result calculate() {
+    final double totalDuration = sourceTotalDuration
+        .andThen(applyExperienceRating)
+        .apply(durationSource);
+
+    final double perTagDuration = applyDurationSplitStrategy.apply(totalDuration);
+
+    return new Result(Math.round(perTagDuration), Math.round(totalDuration));
+  }
+
+  /**
+   * The result of running DurationCalculator#calculate
+   */
+  public static class Result {
+
+    private long perTagDuration;
+    private long totalDuration;
```

Comment:
Does it make sense to round to `long`s at this point? Or will connectors need to perform further calculations on this result?